documents/design/oavb_ipc.asciidoc

```diff
@@ -0,0 +1,83 @@
+= Interprocess Communication Recommendations for AVB/TSN
+Andrew Elder
+v0.1
+
+== Introduction
+
+The AVB/TSN software stack is comprised of a number of modules that implement protocol stacks specified by IEEE standards. Each of these protocols implement an Ethernet network facing interface and an application facing interface. The required application facing interface is described in some detail in http://avnu.org/wp-content/uploads/2014/05/AVnu_SWAPIs_v1.0.pdf[AVB Software Interfaces and Endpoint Architecture Guidelines]. Since a module running a particular protocol may be a standalone execution unit (for example a Linux daemon), an interprocess communication method is required to interface to the execution unit.
+
+For portability reasons it is recommended to separate the implementation of the interprocess communication method as much as possible from the logic of the protocol implementation. This applies to both an application that is interfacing to an AVB/TSN module and the implementation of the AVB/TSN module itself.
+
+For the purposes of this document the following straw-man application interfaces are described:
+
+.Straw-man Interfaces
+[source,c/c++]
+----
+
+void AVBTSN_SetParam(uint32_t param);
+
+----
+
+where the AVBTSN portion of the function name would be replaced by the name of the IEEE AVB/TSN protocol module that is being targeted.
+
+
+== Communication Stack Abstraction
```

Comment:
I think this is a reasonable set of layers, but it might be a good idea to decide whether there's a desire to have implementations be compatible with one another and at which layer they ought to be compatible. I imagine the choice could be made differently from platform-to-platform based on the software vendors in that area and the desire for cross-vendor compatibility.  Also, some layers may or may not be necessary depending on the level of integration between components.

In some situations (and I think it's definitely a "best practice" to implement things this way) a protocol module will define all its upper-layer interfaces first as a natively-callable library interface. Then they would provide a wrapper application that provides the IPC layer for when the module is desired to run as a separate process. And on a similar note, not all IPC layers will require data packing/unpacking, and there's a significant overhead to performing that kind of data transformation. I think the layering interfaces ought to provide a more-or-less 'free' way of skipping layers that aren't required on a given platform/IPC combination. I'm guessing you've thought of this already, but I think it's good to make it explicit. It would be nice if the wrapper applications could be written in a standardized way, relying on the same set of libraries to provide the lower-level layering that is required and allowing the layering to be easily swapped for different platforms.

My main concern here is that we provide guidance that's reasonable for the range of platforms from Autosar-based automotive platforms (which take a very synchronous/time-triggered approach to things and require interfaces to have relatively low abstraction overhead) to Linux and Windows platforms (these can tolerate a reasonable amount of abstraction and interface flexibility) and then back again to standalone speakers or sensors/motor controllers with perhaps no OS at all.